File: src/Web/Client/Services/ProjectsService.cs
using Flurl;
using MudBlazor;
using Shared.Dtos;
using Shared.Enums;
using Shared.ViewModels;

namespace Web.Client.Services;

public class ProjectsService : ApiService
{
    public ProjectsService(HttpClient httpClient, ISnackbar snackbar) 
        : base(httpClient, snackbar)
    {
    }

    public async Task<ProjectsVM?> GetForOrganizationUser(Guid organizationId)
        => await Get<ProjectsVM>($"projects/organization/{organizationId}/user");

    public async Task<bool> Create(Guid organizationId, CreateProjectDto model)
        => await Post($"projects/organization/{organizationId}", model);

    public async Task<UsersSearchVM?> GetAvailableUsers(Guid projectId, Guid organizationId)
    {
        var url = "users/available-for-project"
                .SetQueryParam("projectId", projectId)
                .SetQueryParam("organizationId", organizationId)
                .ToString();
        return await Get<UsersSearchVM>(url);
    }

    public async Task<RolesVM<ProjectPermissions>?> GetRoles(Guid projectId)
        => await Get<RolesVM<ProjectPermissions>>($"projects/{projectId}/roles");

    public async Task<ProjectMembersVM?> GetMembers(Guid projectId)
        => await Get<ProjectMembersVM>($"projects/{projectId}/members");

    public async Task<bool> AddMember(Guid projectId, AddProjectMemberDto model)
        => await Post($"projects/{projectId}/members", model);

    public async Task<bool> RemoveMember(Guid projectId, Guid memberId)
        => await Post($"projects/{projectId}/members/{memberId}/remove");

    public async Task<ProjectOrganizationVM?> GetOrganization(Guid projectId)
        => await Get<ProjectOrganizationVM>($"projects/{projectId}/organization");

    public async Task<ProjectSettingsVM?> GetSettings(Guid projectId)
        => await Get<ProjectSettingsVM>($"projects/{projectId}/settings");

    public async Task<bool> UpdateName(Guid projectId, UpdateProjectNameDto model)
        => await Post($"projects/{projectId}/update-name", model);

    public async Task<bool> Delete(Guid projectId)
        => await Post($"projects/{projectId}/delete");

    public async Task<bool> UpdateMemberRole(Guid projectId, Guid memberId, UpdateMemberRoleDto model)
        => await Post($"projects/{projectId}/members/{memberId}/update-role", model);

    public async Task<bool> UpdateRoleName(Guid projectId, Guid roleId, UpdateRoleNameDto model)
        => await Post($"projects/{projectId}/roles/{roleId}/update-name", model);

    public async Task<bool> UpdateRolePermissions(Guid projectId, Guid roleId, UpdateRolePermissionsDto<ProjectPermissions> model)
        => await Post($"projects/{projectId}/roles/{roleId}/update-permissions", model);

    public async Task<bool> CreateRole(Guid projectId, CreateRoleDto<ProjectPermissions> model)
        => await Post($"projects/{projectId}/roles", model);

    public async Task<bool> DeleteRole(Guid projectId, Guid roleId)
        => await Post($"projects/{projectId}/roles/{roleId}/delete");
}

File: src/Web/Client/composables/projectsService.ts
import type { CreateProjectDto, CreateProjectInvitationDto, UpdateProjectNameDto } from "~/types/dtos/projects";
import type { CreateRoleDto, DeleteRoleDto, RemoveMemberDto, UpdateMemberRoleDto, UpdateRoleNameDto, UpdateRolePermissionsDto } from "~/types/dtos/projects";
import type { RolesVM, ProjectMembersVM, ProjectInfoVM, ProjectsVM, ProjectSettingsVM, UserProjectPermissionsVM, ProjectInvitationsVM, UserProjectInvitationsVM } from "~/types/viewModels/projects";

export const useProjectsService = () => {
    const api = useApi();

    return {
        async getProjects() {
            return await api.sendGetRequest<ProjectsVM>(`projects`);
        },
        async getInfo(id: string) {
            return await api.sendGetRequest<ProjectInfoVM>(`projects/${id}`);
        },
        async createProject(model: CreateProjectDto) {
            await api.sendPostRequest('projects', model);
        },
        async createInvitation(id: string, model: CreateProjectInvitationDto) {
            await api.sendPostRequest(`projects/${id}/invitations`, model);
        },
        async getInvitations(id: string) {
            return await api.sendGetRequest<ProjectInvitationsVM>(`projects/${id}/invitations`);
        },
        async getUserInvitations() {
            return await api.sendGetRequest<UserProjectInvitationsVM>('projects/invitations');
        },
        async acceptInvitation(invitationId: string) {
            await api.sendPostRequest(`projects/invitations/${invitationId}/accept`, undefined);
        },
        async declineInvitation(invitationId: string) {
            await api.sendPostRequest(`projects/invitations/${invitationId}/decline`, undefined);
        },
        async cancelInvitation(id: string, invitationId: string) {
            await api.sendPostRequest(`projects/invitations/${invitationId}/cancel`, undefined,  { 'ProjectId': id });
        },
        async getMembers(id: string) {
            return await api.sendGetRequest<ProjectMembersVM>(`projects/${id}/members`);
        },
        async getRoles(id: string) {
            return await api.sendGetRequest<RolesVM>(`projects/${id}/roles`);
        },
        async updateMemberRole(id: string, model: UpdateMemberRoleDto) {
            await api.sendPostRequest(`projects/${id}/members/role`, model);
        },
        async createRole(id: string, model: CreateRoleDto) {
            await api.sendPostRequest(`projects/${id}/roles`, model);
        },
        async updateRolePermissions(id: string, model: UpdateRolePermissionsDto) {
            await api.sendPostRequest(`projects/${id}/roles/permissions`, model);
        },
        async updateRoleName(id: string, model: UpdateRoleNameDto) {
            await api.sendPostRequest(`projects/${id}/roles/name`, model);
        },
        async deleteRole(id: string, model: DeleteRoleDto) {
            await api.sendPostRequest(`projects/${id}/roles/delete`, model);
        },
        async removeMember(id: string, model: RemoveMemberDto) {
            await api.sendPostRequest(`projects/${id}/members/remove`, model);
        },
        async getSettings(id: string) {
            return await api.sendGetRequest<ProjectSettingsVM>(`projects/${id}/settings`);
        },
        async updateName(id: string, model: UpdateProjectNameDto) {
            await api.sendPostRequest(`projects/${id}/name`, model);
        },
        async deleteProject(id: string) {
            await api.sendPostRequest(`projects/${id}/delete`, undefined);
        },
        async getUserPermissions(id: string) {
            return await api.sendGetRequest<UserProjectPermissionsVM>(`projects/${id}/permissions`);
        },
        async leaveProject(id: string) {
            await api.sendPostRequest(`projects/${id}/leave`, undefined);
        },
    }
}